src/lexer/init.lua

```diff
@@ -263,6 +264,51 @@ function lexer.navigator()
 		end)
 	end
 
+	function nav:HotswapSource(newSource)
+		local oldSource = self.Source or ""
+		self.Source = newSource
+
+		local minimumLength, maximumLength = 0, math.min(#oldSource, #newSource)
+		while minimumLength < maximumLength do
+			local mid = (minimumLength + maximumLength + 1) // 2
+
+			if
+				oldSource:byte(mid) == newSource:byte(mid) -- cheap check of last character
+				and oldSource:sub(1, mid - 1) == newSource:sub(1, mid - 1) -- expensive check of all previous characters
```

Comment:
There's a faster way to do this. Because we already checked that `1 -> minimumLength` is unchanged, we actually only need to check the new `minimumLength -> mid` in each iteration, allowing us to use smaller substrings.

<img width="1034" height="364" alt="image" src="https://github.com/user-attachments/assets/29129711-9a78-4dde-bdf3-4c0c99791277" />



```suggestion
				and oldSource:sub(minimumLength, mid - 1) == newSource:sub(minimumLength, mid - 1) -- expensive check of all previous characters
```